python/taichi/lang/matrix_ops_utils.py

```diff
@@ -50,9 +76,64 @@ def assert_tensor(m, msg='not tensor type: {}'):
     raise TaichiCompilationError(msg.format(type(m)))
 
 
+def assert_vector(v, msg='not a vector: {}'):
+    if (isinstance(v, Expr) or isinstance(v, Matrix)) and len(
+            v.get_shape()) == 1:
+        return True, None
+    raise TaichiCompilationError(msg.format(type(v)))
+
+
+def assert_list(x, msg='not a list: {}'):
+    if isinstance(x, list):
+        return True, None
+    raise TaichiCompilationError(msg.format(type(x)))
+
+
+def same_shapes(xs):
+    shapes = [x.get_shape() for x in xs]
+    if len(set(shapes)) != 1:
+        return False, f'required shapes to be the same, got shapes {shapes}'
+    return True, None
+
+
 def square_matrix(x):
     assert_tensor(x)
     shape = x.get_shape()
     if shape[0] != shape[1]:
         return False, f'not a square matrix: {shape}'
     return True, None
+
+
+def dim_lt(dim, limit, msg=None):
+    def check(x):
+        assert_tensor(x)
+        shape = x.get_shape()
+        return shape[dim] < limit, (
+            f'Dimension >= {limit} is not supported: {shape}'
+            if not msg else msg.format(shape))
+
+    return check
+
+
+def is_int_const(x):
+    if isinstance(x, int):
+        return True, None
+    if isinstance(x, Expr) and x.val_int() is not None:
+        return True, None
+    return False, f'not an integer: {x} of type {type(x).__name__}'
+
+
+def check_matmul(x, y):
+    assert_tensor(x, f'left hand side is not a matrix: {type(x)}')
+    assert_tensor(y, f'right hand side is not a matrix: {type(y)}')
+    x_shape = x.get_shape()
+    y_shape = y.get_shape()
+    if len(x_shape) == 1:
+        if len(y_shape) == 1:
+            return True, None
+        if x_shape[0] != y_shape[1]:
+            return False, f'dimension mismatch between {x_shape} and {y_shape} for left multiplication'
+    else:
+        if x_shape[0] != y_shape[0]:
```

Comment:
```suggestion
        if x_shape[1] != y_shape[0]:
```
`x_shape[1]` is the number of columns of `x`.